### ISSUE TITLE
Update `SettingsNotice` component to account for new GA4 notice design

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3251]
#3248, #3249, and #3250 implemented the UI for the new Analytics setup flow with GA4 support. The original [Figma design](https://www.figma.com/file/ix0pcdTQi1Bf9QFNBlBByB/GA4-%2B-UA-onboarding?node-id=20269%3A8621) includes a new visual notice look that those issues didn't consider yet. This issue here is about implementing it.

**While this is technically blocked by the above issues, the majority of work needed here can be started without waiting for these issues.**

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The existing `SettingsNotice` component should be adjusted:
    * It should support an additional optional `Icon` prop which should expect a `WPComponent`, typically an SVG. If passed, it should be rendered on the left of the notice in its own "column" with reserved space for it. The current default star that is added to the notice using CSS should be removed (i.e. the `Icon` prop should serve as more flexible replacement for this approach).
    * It should support an additional optional `LearnMore` prop which should expect a `WPComponent`, typically a `Link` that provides more information. If passed, it should be rendered on the (bottom) right of the notice. The mention of "bottom" here is important for the case that the main content of the notice (i.e. `children`) is taller than one line.
    * See [design for a single-line notice](https://www.figma.com/file/ix0pcdTQi1Bf9QFNBlBByB/GA4-%2B-UA-onboarding?node-id=20269%3A8621) and [design for a multi-line notice](https://www.figma.com/file/ix0pcdTQi1Bf9QFNBlBByB/GA4-%2B-UA-onboarding?node-id=20269%3A7786).
* The new `SettingsNotice` props should be used in the component usages in `GA4Notice`, `SetupFormUA`, `SetupFormGA4`, and `SetupFormGA4Transitional`:
    * In each case, the `Icon` prop should be the star icon from the design, and the `LearnMore` prop should be the respective learn more link.

## Implementation Brief

* Edit the `googlesitekit-settings-notice__text::before` SASS and remove the background images
* Edit the `SettingsNotice` component
* Add an optional `Icon` prop and use grid components to render a 2 column grid to output the icon and text side by side and remove the padding to the left of the text
* Add an optional `LearnMore` prop which is outputted in the bottom right of the notice below the text.  (see designs referenced in the AC)
* In the the following components pass the relevant icon using the new `Icon` prop and the learn more link using the new `LearnMore` prop:
  * `GA4Notice`
  * `SetupFormUA`
  * `SetupFormGA4`
  * `SetupFormGA4Transitional`

[Branch containing progress](https://github.com/google/site-kit-wp/tree/enhancement/3251-settings-notice)

### Test Coverage

* N/A

### Visual Regression Changes

* check visual regression tests and update as necessary

## QA Brief

Check that:
* the three setup forms (as mentioned in the AC) look like the design https://google.github.io/site-kit-wp/storybook/develop/?path=/story/modules-analytics-setup-setupformga4--ready
* the GA4Notice looks as per the designs https://google.github.io/site-kit-wp/storybook/develop/?path=/story/analytics-module--ga4-notice
* the Notices stories look ok https://google.github.io/site-kit-wp/storybook/develop/?path=/story/global-notices--settings-warning-notice

## Changelog entry

* N/A

**Comment by aaemnnosttv** [comment]
> * Add an optional `icon` prop [...]
> * Add an optional 'learnMore' prop [...]

@Hazlitte note that these props are intentionally capitalized as `Icon` and `LearnMore` respectively. This is a convention we have to indicate that the value is a component type rather than an element (or some other primitive value).

> output it in a new `div` before the text `div` with `inline-block` styling so they appear side by side

We shouldn't render an icon inline like this with text or other inline elements because it will wrap oddly if too long. Instead, we should probably use grid components to render a 2 column grid (one for the icon, and everything else in the other). Note in the designs that the "body" of the notice does not wrap below the icon, so that would be in its own column (grid or not). Also, the current `googlesitekit-settings-notice__text` styles has extra left padding which was to allow space for the background image before so this should also be removed.

**Comment by aaemnnosttv** [comment]
> Note that at the time of writing the IB only the GA4Notice component mentioned in the AC is present in the develop branch. When executing check for the existence of the SetupFormUA, SetupFormGA4 and SetupFormGA4Transitional components and update them as necessary to work with the new SettingsNotice props.

This part should be unnecessary since this issue depends on the issues that implement these, so they should exist by the time this can be worked on.

IB ✅

**Comment by Hazlitte** [comment]
@aaemnnosttv Would you like me to remove that paragraph from the IB?

**Comment by aaemnnosttv** [comment]
@Hazlitte yeah, let's update it to explicitly mention updating these (as if they already exist). As it is, it might be interpreted as more of an optional detail rather than a requirement as mentioned in the ACs.

**Comment by Hazlitte** [comment]
@aaemnnosttv Ok, I have updated the IB accordingly 👍

**Comment by aaemnnosttv** [comment]
Thanks @Hazlitte !

IB ✅

**Comment by Hazlitte** [comment]
[Branch containing progress](https://github.com/google/site-kit-wp/tree/enhancement/3251-settings-notice)

**Comment by danielgent** [comment]
There's no mention of the story for `<SettingsNotice />`  :thinking: 
https://google.github.io/site-kit-wp/storybook/pull/3330/?path=/story/global-notices--settings-warning-notice


This shows `SettingsNotice` accepting a `type` prop which is an enum
```
	.add( 'Settings warning notice', () => (
		<SettingsNotice type={ TYPE_WARNING }>
			{ 'This is a warning.' }
		</SettingsNotice>
	) )
	.add( 'Settings info notice', () => (
		<SettingsNotice type={ TYPE_INFO }>
			{ 'This is an information.' }
		</SettingsNotice>
	) )
	.add( 'Settings suggestion notice', () => (
		<SettingsNotice type={ TYPE_SUGGESTION }>
			{ 'This is a suggestion.' }
		</SettingsNotice>
	) );
```

Reading the AC I would assume that we would get rid of this prop... and that the background will always be blue, and that the icon is whatever is passed in :thinking: 

In which case this story should be changed to have
* SettingsNotice with Icon and LearnMore props
* SettingsNotice with Icon prop
* SettingsNotice with LearnMore prop
* SettingsNotice with only children prop

*Also* I'm not clear if the props should be functions or nodes :thinking:  (should they be instantiated Components?)
I would expect them to be nodes e.g. 
```
<SettingsNotice Icon={<Icon/>} LearnMore={<LearnMore/>}>
  I am a notice
</SettingsNotice/>
```

As opposed to
```
<SettingsNotice Icon={Icon} LearnMore={LearnMore}>
  I am a notice
</SettingsNotice/>
```

I don't think _"components, elements, and instances"_ ever became accepted generally as terminology :stuck_out_tongue: 
https://reactjs.org/blog/2015/12/18/react-components-elements-and-instances.html

@aaemnnosttv

**Comment by danielgent** [comment]
Something else to check

According to the design, the `children` is responsible for not flowing over `LearnMore` e.g.
![design overflow](https://user-images.githubusercontent.com/8496063/119512230-69de5800-bd73-11eb-8b80-6ed29f2f84d7.png)

No problem for text like this
![notice-looks-fine](https://user-images.githubusercontent.com/8496063/119512283-75318380-bd73-11eb-98ab-48065b717bd8.png)


But is this expected? :thinking: 
![multiline not ok](https://user-images.githubusercontent.com/8496063/119512320-7e225500-bd73-11eb-8b32-f1086618c12d.png)

**Comment by danielgent** [comment]
Progress so far can be seen here on storybook https://google.github.io/site-kit-wp/storybook/pull/3404/?path=/story/global-notices--settings-info-notice-no-icon

There is no design for a `SettingsNotice` without an `Icon` prop, so I've made the text flush left in this case (easily reverted!)
![flush-left-no-icon](https://user-images.githubusercontent.com/8496063/119519469-bcbb0e00-bd79-11eb-9f5e-13fa56b71c17.png)

**Comment by aaemnnosttv** [comment]
> There's no mention of the story for `<SettingsNotice />` 🤔
> [google.github.io/site-kit-wp/storybook/pull/3330/?path=/story/global-notices--settings-warning-notice](https://google.github.io/site-kit-wp/storybook/pull/3330/?path=/story/global-notices--settings-warning-notice)
> 
> This shows `SettingsNotice` accepting a `type` prop which is an enum
> 
> Reading the AC I would assume that we would get rid of this prop... and that the background will always be blue, and that the icon is whatever is passed in 🤔

@danielgent `type` is still relevant for controlling the background and text colors. Given that `Icon` is an optional prop, `type` should probably also determine the default `Icon` used.

> _Also_ I'm not clear if the props should be functions or nodes 🤔 (should they be instantiated Components?)
> I would expect them to be nodes e.g. 
> I don't think _"components, elements, and instances"_ ever became accepted generally as terminology 😛
> [reactjs.org/blog/2015/12/18/react-components-elements-and-instances.html](https://reactjs.org/blog/2015/12/18/react-components-elements-and-instances.html)

We probably need to document this but WP has its own types on top of React's so if you're not familiar:

- [`WPElement`](https://github.com/WordPress/gutenberg/blob/62280f930fa95ba40f9d121b93e22a3e8f8884d2/packages/element/src/react.js#L31-L35) / `ReactElement` / `PropTypes.element` / `<MyComponent />` – "instantiated component"
- [`WPComponent`](https://github.com/WordPress/gutenberg/blob/62280f930fa95ba40f9d121b93e22a3e8f8884d2/packages/element/src/react.js#L37-L41) / `ComponentType` / `PropTypes.elementType` / `MyComponent` – "uninstantiated component"

More on this:
* https://github.com/google/site-kit-wp/issues/2413
* https://github.com/google/site-kit-wp/issues/2381#issuecomment-730196523

> There is no design for a `SettingsNotice` without an `Icon` prop, so I've made the text flush left in this case (easily reverted!)

@felixarntz may need to clarify this but my understanding is that only the default icon that is added via CSS is to be removed, not the icon entirely. See above, but I think that we can preserve all the same default icons for each type if `Icon` isn't provided.

> According to the design, and following the AC, the `children` is responsible for not flowing over `LearnMore`
> No problem for text like this [...]
> But is this expected?

I don't think so! Not sure what the best solution here is without resorting to old school floats 😛  (thoughts @asvinb ?)

As an aside, I think the grid may not work here as a full column is probably too much for the icon
![image](https://user-images.githubusercontent.com/1621608/119634146-19bad080-be1b-11eb-9703-abea791afb68.png)

Seems like using flexbox might be the better solution here 🤔

**Comment by aaemnnosttv** [comment]
> `Given that Icon is an optional prop, type should probably also determine the default Icon used.`
> Yes so it says optional prop but doesn't mention defaults
> IMHO it should set the Icon based on the type, allowing overrides by the Icon prop.
> Should I just do that? Does it need adding to the IB? Should this go back to IBR? what's the process?

@danielgent Issues very rarely ever go back to IB (+IBR) unless they need a completely different approach (I think I can count the times this happened on one finger). In this case, it's more of a detail that wasn't explicitly mentioned in the IB, but that's okay – it's not uncommon for things to come up during implementation that weren't thought of ahead of time. The defaults are already in place, the difference is that they weren't flexible before, so this issue is about allowing for flexibility without breaking the way it's currently used.

> `seems like using flexbox might be the better solution here`
> Agreed! but the IB says grid (which is what @Hazlitte implemented as expected!)

It's good to ask before deviating too far from the IB but the IB is the intended overall approach; it's not set in stone. The ACs are the part which needs to always be fulfilled (although these can be flexible/amended when needed). We should always prefer to use the Material grid for grid-like layouts over something more custom but we're not forced to use it. The IB was correct to outline this as the approach but if it's not going to work out and we have valid reasons for not using it then it's perfectly fine to go with a flexbox-based layout.

**Comment by danielgent** [comment]
@felixarntz  We're going to wait until your back to make a decision on what to do here

The gist of it is that there are two layouts for the `SettingsNotice` component

The first is like one row of flex
![flex](https://user-images.githubusercontent.com/8496063/120526074-1ace9d80-c3d9-11eb-866b-bc060108e676.png)

The second is like two rows both flex
![example](https://user-images.githubusercontent.com/8496063/120526078-1b673400-c3d9-11eb-884d-1bf6e2b4c9c2.png)

There's no way we can have one `children` prop that is then full-width on the first line if the `children` is more than one line :smile:  _(well not without some hacky DOM measuring code which will be complex especially when going mobile)_ 

So we can have two options
1. Have two components - e.g. `SettingsNotice` and `SettingsNoticeMultiline`.  The latter can then have two props e.g. `notice` and `children` as the `GA4PropertyNotice` has
2. Have just one component that takes `notice` and `children` and then renders only one row if `children` is empty

**Comment by felixarntz** [comment]
@danielgent Good catch! I suggest then going with your second approach - essentially `children` then is something like "additional children", other than the notice itself. That sounds good to me.

**Comment by aaemnnosttv** [comment]
## QA ⚠️ 

While reviewing the components I noticed a few things that were off, but also worth noting that the designs do not show the appearance of all types of notices.

I think we probably need a follow-up issue here to address some of it though as it's beyond the scope of this issue.

I created a follow-up PR to address these things here: https://github.com/google/site-kit-wp/pull/3533

**Comment by wpdarren** [comment]
### QA Update:  Pass ✅ 

- Went through the setup screens on Storyboard to Figma designs. There are some slight differences but these were brought up in the bug bash. 
- The GA4Notice looks as per the designs.
- The notices look okay. There are some alignment and spacing issues but this was brought up in the bug bash. 

![image](https://user-images.githubusercontent.com/73545194/121555328-d2336780-ca0a-11eb-9050-1ffece674f5f.png)
